### PR TITLE
105 remove backend code

### LINKED ISSUE #105: Remove getVisualization webservice and replace with frontend implementation
The plugin currently depends on the custom backend webservice getVisualization. This is the last leftover from "calc and layout in the backend" to "all logic in the frontend" thing we do for a while now.

At the end, there should be no more getVisualization service.

### PR DESCRIPTION
**Pull request by stefanrinderle**
This is a lot of code, but it's also a lot of functionality. 

This feature does not close #105  but it's a first working version. I'm sure that we find something within the review, which needs to be done ;-)

### CODE REVIEW

File: softvis3d-frontend/src/services/sonarqube/measures/SonarQubeMeasureResponse.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+/**
+ * BRC - Sub-projects
+ * DIR - Directories
+ * FIL - Files
+ * TRK - Projects
+ * UTS - Unit Test Files
+ */
+export type SonarQubeQualifier =
```
Comment (Ungolianth) [inline softvis3d-frontend/src/services/sonarqube/measures/SonarQubeMeasureResponse.ts]: Turn these into meaningful reusable Constants in the form of human readable Code instead of "Magic Constants" with a comment

File: softvis3d-frontend/src/services/sonarqube/SonarQubeTransformer.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+            if (localMeasure.value) {
+                measureList[localMeasure.metric] = +localMeasure.value;
+            } else {
+                if (localMeasure.periods.length > 0) {
+                    measureList[localMeasure.metric] = +localMeasure.periods[0].value;
+                }
+            }
+        }
+
+        let isFile: boolean = component.qualifier === "FIL" || component.qualifier === "UTS";
```
Comment (Ungolianth) [inline softvis3d-frontend/src/services/sonarqube/SonarQubeTransformer.ts]: "Magic Constants" (see above)

File: softvis3d-frontend/src/stores/SceneStore.ts [hunk 3]
```diff
@@ -31,6 +32,8 @@ class SceneStore {
     public get selectedElement(): TreeElement | null {
         let selectedElement: TreeElement | null = null;
         if (this.legacyData !== null && this.selectedObjectId != null) {
+            // TODO: check performance of searchTreeNode on bigger projects
```
Comment (Ungolianth) [inline softvis3d-frontend/src/stores/SceneStore.ts]: Do you want to resolve this before merging to master?

File: softvis3d-frontend/src/services/sonarqube/measures/SonarQubeMeasuresService.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        this.sceneStore = sceneStore;
+    }
+
+    public loadMeasuresInitial(options: VisualizationOptions) {
+        this.appStatusStore.load(SonarQubeMeasuresService.LOAD_MEASURES);
+
+        this.sceneStore.options = options;
+        this.sceneStore.shapes = null;
+
+        // this.projectKey = "org.sonarsource.sonarqube:sonarqube";
```
Comment (Ungolianth) [inline softvis3d-frontend/src/services/sonarqube/measures/SonarQubeMeasuresService.ts]: 🔥

File: softvis3d-frontend/src/services/sonarqube/measures/SonarQubeMeasuresTreeService.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+    public loadTree(parent: TreeElement, metricKeys: string): Promise<void> {
+        this.appStatusStore.loadStatusUpdateIncrementMax(SonarQubeMeasuresService.LOAD_MEASURES);
+
+        return new Promise<void>((resolve, reject) => {
+            /**
+             * Load the direct children of the given component. In SQ terms this means only directories or sub-projects
+             * will be loaded.
+             */
+            this.measureApiService.loadMeasures(parent.key, metricKeys, "children", ["DIR", "BRC"]).then((result) => {
```
Comment (Ungolianth) [inline softvis3d-frontend/src/services/sonarqube/measures/SonarQubeMeasuresTreeService.ts]: More Magic Constants

File: softvis3d-frontend/src/classes/TreeElement.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        this.path = path;
+        this.isFile = isFile;
+        this.parent = parent;
+        this.children = [];
+    }
+
+    /**
+     * Will be called with the path of the components sorted.
+     */
+    public addAsChild(element: TreeElement, isDescending: boolean = false) {
```
Comment (stefanrinderle) [inline softvis3d-frontend/src/classes/TreeElement.ts]: Yes, would you rename it to just add?

Comment (Ungolianth) [inline softvis3d-frontend/src/classes/TreeElement.ts]: Unless you plan on adding a different algorithm to `add` an item I would rename it, yes. But this is just a cosmetic advice anyway.

File: softvis3d-frontend/src/classes/TreeElement.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        let children = this.children;
+        if (isDescending) {
+            children = children.reverse();
```
Comment (stefanrinderle) [inline softvis3d-frontend/src/classes/TreeElement.ts]: The base idea of isDescending is the order in which the child elements will be checked. It was meant to be faster to search for the matching child element in a reverse oder.

Fixed the code to reverse a real copy of the children instead of the array itself.

Comment (Ungolianth) [inline softvis3d-frontend/src/classes/TreeElement.ts]: So renaming the parameter to `examineInReversedOrder` would actually be more fitting?

Also: Does it make sense then to just get rid of the parameter and change the loop to always search in reversed order?

File: softvis3d-frontend/src/classes/TreeElement.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        }
+
+        for (let child of children) {
```
Comment (stefanrinderle) [inline softvis3d-frontend/src/classes/TreeElement.ts]: `if (this.path === element.path) {`

Does never work (at least with the SQ data) as every element in the tree has a unique path and every element is only added once. this.path === element.path would never be true. The question here would be "is it inside the current node or should it be deeper in the tree" which could be something like:

`let remainingPath: string = element.path.substr(this.path.length + 1, element.path.length);
        if (remainingPath.indexOf("/") !== 0) {...`

I tried it with the above code but it did not work as expected. I would like to leave it as is.

File: softvis3d-frontend/src/classes/TreeElement.ts (hunk 7 above)
Comment (stefanrinderle) [inline softvis3d-frontend/src/classes/TreeElement.ts]: Parameter renaming done.

For adding directories, ascending order is better, for files descending.

Comment (Ungolianth) [inline softvis3d-frontend/src/classes/TreeElement.ts]: I see. Since the algorithm does nothing close to an ordered insert this behavior would ascribe to the input data (SonarQube) though and is not universally applicable, right?

File: softvis3d-frontend/src/classes/status/LoadAction.ts [hunk 9]
```diff
@@ -14,16 +14,32 @@ export default class LoadAction extends StatusAction {
         this._current = current;
     }
 
-    public get limit(): number {
+    public get limit(): number | undefined {
```
Comment (stefanrinderle) [inline softvis3d-frontend/src/classes/status/LoadAction.ts]: The thing is that the limit can be undefined. There are LoadAction which defined a limit and ones which did not. In addition, the limit can be changed via "incrementMax". This is needed by the asynchronous recursive logic where it is not clear in the beginning how many requests it has to do.

But you're right, the resolve of undefined in all the classes is not so nice. I created a method hasLimits() as follows:

`public hasLimit(): boolean {
        return typeof this.current !== "undefined" && typeof this.limit !== "undefined";
}`

Unfortunately, the compile creates an undefined exception also the above method was called:

`if (this.hasLimit()) {
            this._limit = this._limit + 1;
 }`

"Object is possibly 'undefined'." Any ideas how to solve this?

File: softvis3d-frontend/src/classes/TreeElement.ts (hunk 8 above)
Comment (Ungolianth) [inline softvis3d-frontend/src/classes/TreeElement.ts]: To bad. But as long as we don't have to worry about the performance this is not an issue. 

✅ Discussion resolved

File: softvis3d-frontend/src/classes/status/LoadAction.ts (hunk 9 above)
Comment (Ungolianth) [inline softvis3d-frontend/src/classes/status/LoadAction.ts]: So in every scenario a LoadActions "limit" should be increment-able?

Then you can initialise the value with `0` in the constructor. the value is never `undefined`.

File: softvis3d-frontend/src/classes/status/LoadAction.ts [hunk 10]
```diff
@@ -14,16 +14,32 @@ export default class LoadAction extends StatusAction {
...
         return this._limit;
     }
 
-    public get current(): number {
+    public get current(): number | undefined  {
         return this._current;
     }
 
+    public incrementMax() {
```
Comment (stefanrinderle) [inline softvis3d-frontend/src/classes/status/LoadAction.ts]: Thanks. Renamed to max and current.

File: softvis3d-frontend/src/classes/TreeElement.ts (hunk 6 above)
Comment (stefanrinderle) [inline softvis3d-frontend/src/classes/TreeElement.ts]: Renamed to add

File: softvis3d-frontend/src/services/sonarqube/measures/SonarQubeMeasureResponse.ts (hunk 1 above)
Comment (stefanrinderle) [inline softvis3d-frontend/src/services/sonarqube/measures/SonarQubeMeasureResponse.ts]: The constant names come from the SQ api (see qualifier here http://softvis3d.com/sonar/web_api/api/measures). As i didn't wanted to use it as string values, i created this type. I tried something like this:

`export const subProjectQualifier: string = "BRC";

export type SonarQubeQualifier =
    subProjectQualifier |...`

but i does not work. Any ideas?

File: softvis3d-frontend/src/services/sonarqube/measures/SonarQubeMeasuresService.ts (hunk 4 above)
Comment (stefanrinderle) [inline softvis3d-frontend/src/services/sonarqube/measures/SonarQubeMeasuresService.ts]: ups

File: softvis3d-frontend/src/stores/SceneStore.ts (hunk 3 above)
Comment (stefanrinderle) [inline softvis3d-frontend/src/stores/SceneStore.ts]: checked it. Pretty fast - about 1 millisecond

File: softvis3d-frontend/src/services/sonarqube/measures/SonarQubeMeasureResponse.ts (hunk 1 above)
Comment (Ungolianth) [inline softvis3d-frontend/src/services/sonarqube/measures/SonarQubeMeasureResponse.ts]: Option 1: Make them part of a class
```TS

export class SonarQubeApi {
    public static TYPE_SUBPROJECT = "BRC";
    public static TYPE_DIRECTORY = "DIR";
    // ....
}
```
------------
Option 2: Standalone Constants to be imported
```TS

const TYPE_SUBPROJECT = "BRC";
const TYPE_DIRECTORY = "DIR";
// ....

export {
    TYPE_DIRECTORY,
    TYPE_SUBPROJECT
    //...
}
```

File: softvis3d-frontend/src/components/sidebar/FolderContent.tsx [hunk 11]
```diff
@@ -77,20 +78,23 @@ export default class FolderContent extends React.Component<NodeListProps, NodeLi
...
-        for (let child of folder.children) {
-            folderElements.push(
-                <FolderContentElement
-                    key={child.id}
-                    element={child}
-                    isSelected={child.id === sceneStore.selectedObjectId}
-                    sceneStore={sceneStore}
-                />
-            );
+        for (let child of folder.children.filter((e) => !e.isFile)) {
```
Comment (stefanrinderle) [inline softvis3d-frontend/src/components/sidebar/FolderContent.tsx]: Moved sorting code to TreeElement class which reduces the logic in FolderContent. Nice sort function. As a java developer, this is no really easier to understand ;-) but i get used to it slowly.

File: softvis3d-frontend/src/classes/status/LoadAction.ts (hunk 9 above)
Comment (stefanrinderle) [inline softvis3d-frontend/src/classes/status/LoadAction.ts]: The thing for me here is that it can be undefined in the "real world". So our code has to handle this and it has to be checked in the code. 

If i would set max and current to 0 i am not longer able to know if the progress bar should be visible or not - or the code would check for 0 as value. With undefined, this is clear - at least for me ;-).

I cleaned up a bit, so maybe this one is more clear now.

File: softvis3d-frontend/src/classes/status/StatusActionQueue.ts [hunk 12]
```diff
@@ -9,6 +9,16 @@ export default class StatusActionQueue<T extends StatusAction> {
         this._queue.push(action);
     }
 
+    public copyAndRemove(action: T): StatusActionQueue<T> {
```
Comment (stefanrinderle) [inline softvis3d-frontend/src/classes/status/StatusActionQueue.ts]: I checked the implementation again and a new copy of the StatusActionQueue is only needed on status updates. These methods now return a new instance. What do you think?

File: softvis3d-frontend/src/classes/TreeElement.ts (hunk 7 above)
Comment (stefanrinderle) [inline softvis3d-frontend/src/classes/TreeElement.ts]: Rigth, moved the method to the SonarQubeTransformer as is contains SQ specific code.

File: softvis3d-frontend/src/services/sonarqube/measures/SonarQubeMeasureResponse.ts (hunk 1 above)
Comment (stefanrinderle) [inline softvis3d-frontend/src/services/sonarqube/measures/SonarQubeMeasureResponse.ts]: Created and used the constants.

File: softvis3d-frontend/src/classes/status/StatusActionQueue.ts [hunk 13]
```diff
@@ -21,7 +22,13 @@ export default class StatusActionQueue<T extends StatusAction> {
...
     }
 
-    public copyAndUpdate(action: T): StatusActionQueue<T> {
+    /**
+     * As only the content of the action changes, queue has to be recreated
+     * for the @observable to fire an event.
+     *
+     * @returns new instance of StatusActionQueue<T>.
+     */
+    public update(action: T): StatusActionQueue<T> {
```
Comment (Ungolianth) [inline softvis3d-frontend/src/classes/status/StatusActionQueue.ts]: I hope you don't mind me pulling this peace of code back on the table! I just tried to update this, but realised after more than an hour that the issue lies much deeper than the comment suggests:

 * An Update (eg. increaseMax) is triggered
 * The AppStatusStore fetches the Action and Updates the value. _This does not generate a new Object!_
 * The AppStatusStore then asks the Queue to overwrite the action with the same(!) element. `this._queue[i] = sameElement` would not trigger a change in mobx because actually nothing changes.
 * Therefore the complete array is duplicated and then replaced, just to trigger the change.

--------------------------------

I feel this is actually to much compensation and bad/opaque design just to "make it work" with MobX. 
 * Can we use another style/method to display our data in MobX-Conform way?
 * Should we redesign the way LoadActions are stored/managed?

Comment (stefanrinderle) [inline softvis3d-frontend/src/classes/status/StatusActionQueue.ts]: I'm open to everything here. If you have an idea, give it a try. But i think this should be done in a different pull request or event an different issue ticket then.
